readium/navigator/src/main/java/org/readium/r2/navigator/media2/MediaSessionNavigator.kt

```diff
@@ -0,0 +1,274 @@
+                state = state,
+                rate = playbackSpeed.toDouble(),
+                currentIndex = currentItem.index,
+                currentLink = publication.readingOrder[currentItem.index],
+                currentPosition = currentPosition,
+                bufferedPosition = bufferedPosition
+            )
+        }
+
+    override val playback: StateFlow<MediaNavigator.Playback> =
+        playbackStateFlow.stateInFirst(coroutineScope)
+
+    override suspend fun setPlaybackRate(rate: Double): Try<Unit, MediaNavigator.Exception> =
+        playerFacade.setPlaybackSpeed(rate).toNavigatorResult()
+
+    override suspend fun play(): Try<Unit, MediaNavigator.Exception> =
+        playerFacade.play().toNavigatorResult()
+
+    override suspend fun pause(): Try<Unit, MediaNavigator.Exception> =
+        playerFacade.pause().toNavigatorResult()
+
+    override suspend fun seek(index: Int, position: Duration): Try<Unit, MediaNavigator.Exception> =
+        playerFacade.seekTo(index, position).toNavigatorResult()
+
+    override suspend fun go(locator: Locator): Try<Unit, MediaNavigator.Exception> {
+        Timber.d("Go to locator $locator")
+        val itemIndex = checkNotNull(publication.readingOrder.indexOfFirstWithHref(locator.href))
+        val position = locator.locations.time ?: Duration.ZERO
+        return seek(itemIndex, position)
+    }
+
+    override suspend fun go(link: Link) =
+        go(link.toLocator())
+
+    override suspend fun goForward(): Try<Unit, MediaNavigator.Exception> =
+        seekBy(configuration.skipForwardInterval)
+
+    override suspend fun goBackward(): Try<Unit, MediaNavigator.Exception> =
+        seekBy(-configuration.skipBackwardInterval)
+
+    private suspend fun seekBy(offset: Duration): Try<Unit, MediaNavigator.Exception> =
+        this.playerFacade.playlist!!.durations
+            ?.let { smartSeekBy(offset, it) }
+            ?: dummySeekBy(offset)
+
+    private suspend fun smartSeekBy(
+        offset: Duration,
+        durations: List<Duration>
+    ): Try<Unit, MediaNavigator.Exception> {
+        val (newIndex, newPosition) =
+            SmartSeeker.dispatchSeek(
+                offset,
+                playerFacade.currentPosition!!,
+                playerFacade.currentIndex!!,
+                durations
+            )
+        Timber.d("Smart seeking by $offset resolved to item $newIndex position $newPosition")
+        return playerFacade.seekTo(newIndex, newPosition).toNavigatorResult()
+    }
+
+    private suspend fun dummySeekBy(offset: Duration): Try<Unit, MediaNavigator.Exception> {
+        val newIndex = playerFacade.currentIndex!!
+        val newPosition = playerFacade.currentPosition!! + offset
+        return playerFacade.seekTo(newIndex, newPosition).toNavigatorResult()
+    }
+
+    override fun close() {
+        playerFacade.unregisterPlayerCallback(playerCallback)
+        playerCallback.close()
+        playerFacade.close()
```

Comment:
We probably need to cancel the coroutine scope there too?

```suggestion
        playerFacade.close()
        coroutineScope.cancel()
```

Reply:
Mmm, yes. We often wait for garbage collection, but since there is a `close` method...